src/leo_gateway_s3_api.erl

```diff
@@ -508,8 +532,142 @@ handle_1(Req, [{NumOfMinLayers, NumOfMaxLayers}, HasInnerCache, CustomHeaderSett
+            {Bin, Remains} = extract_atleast(IOList, 80 + 2),
+            <<"chunk-signature=", ChunkSign:64/binary, "\r\n", Rest/binary>> = Bin,
+%            ?debug("aws_chunk_decode", "Chunk Sign: ~p", [ChunkSign]),
+            aws_chunk_decode({ok, Acc}, [Rest, Remains], read_chunk, 0, SignParams#aws_chunk_sign_params{chunk_sign = ChunkSign});
+        _ ->
+            {{ok, Acc}, {IOList, wait_head, 0, SignParams}}
+    end;
+aws_chunk_decode({ok, Acc}, IOList, read_chunk, Offset, #aws_chunk_sign_params{
+                                                  sign_head  = SignHead, 
+                                                  sign_key   = SignKey, 
+                                                  prev_sign  = PrevSign,
+                                                  chunk_sign = ChunkSign,
+                                                  chunk_size = ChunkSize,
+                                                  hash_context = Context
+                                                 } = SignParams) ->
+    ChunkRemainSize = ChunkSize - Offset,
+    case erlang:iolist_size(IOList) of
+        Len when Len >= ChunkRemainSize + 2 ->
+            {Bin, Remains} = extract_atleast(IOList, ChunkRemainSize + 2),
+            <<ChunkPart:ChunkRemainSize/binary, "\r\n", Rest/binary>> = Bin,
+            Context_2 = crypto:hash_update(Context, ChunkPart),
+            ChunkHash = crypto:hash_final(Context_2),
+            ChunkHashBin = leo_hex:binary_to_hexbin(ChunkHash),
+            BinToSign = <<"AWS4-HMAC-SHA256-PAYLOAD\n",
+                          SignHead/binary,
+                          PrevSign/binary,  "\n",
+                          "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\n",
+                          ChunkHashBin/binary>>,
+            Signature = crypto:hmac(sha256, SignKey, BinToSign),
+            Sign = leo_hex:binary_to_hexbin(Signature),
+            case Sign of
+                ChunkSign ->
+                    ?debug("aws_chunk_decode/4", "Output Chunk Size: ~p, Sign: ~p", [ChunkSize, ChunkSign]),
+                    aws_chunk_decode({ok, [Acc, ChunkPart]}, [Rest, Remains], wait_size, 0, SignParams#aws_chunk_sign_params{prev_sign = ChunkSign,
+                                                                                                                             chunk_sign = <<>>});
+                WrongSign ->
+                    ?debug("aws_chunk_decode/4", "BinToSign: ~p", [binary_to_list(BinToSign)]),
+                    ?error("aws_chunk_decode/4", "Chunk Signature Not Match: ~p, ~p", [WrongSign, ChunkSign]),
+                    {{error, unmatch}, {Bin, error, Offset, SignParams}}
+            end;
+        _ ->
+            Bin = iolist_to_binary(IOList),
+            Context_2 = crypto:hash_update(Context, Bin),
+            {{ok, [Acc, Bin]}, {<<>>, read_chunk, Offset + byte_size(Bin), SignParams#aws_chunk_sign_params{hash_context = Context_2}}}
+    end.
 
```

Comment:
It was not a good testing as I only used `aws-sdk-java` to upload 50MB for 5 trials
iolist avg: 13ms (9~20), bin avg: 16ms (12~25)
The result is not significant enough, to be honest.
I plan to try with a larger test size (500MB) after completing the unit test for `leo_s3_libs`
(It would be better to pinpoint `aws_chunk_decode` instead of the whole flow including disk I/O)